src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryQueryBuilder.java

```diff
@@ -0,0 +1,92 @@
+package com.commercetools.sync.inventories.helpers;
+
+import static java.util.Collections.emptyList;
+
+import io.sphere.sdk.inventory.InventoryEntry;
+import io.sphere.sdk.inventory.queries.InventoryEntryQuery;
+import io.sphere.sdk.inventory.queries.InventoryEntryQueryModel;
+import io.sphere.sdk.queries.QueryPredicate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+
+public final class InventoryEntryQueryBuilder {
+
+  /*
+   * The query predicate will be limited to max 10.000 characters(also considered some
+   * conservative space for headers). Because above this size it could return - Error 414 (Request-URI Too Large).
+   * With that one inventory entry could fetch maximum 238 (10000/42 - assumed the sku will be 1 character)
+   * inventory entry.
+   */
+  private static final int MAX_QUERY_LENGTH = 10000;
+
+  /**
+   * Builds {@link List} of {@link InventoryEntryQuery}'s to be used to query existing {@link
+   * InventoryEntry}'s against set of sku and supply channels.
+   *
+   * <p>For instance, a query for one inventory entry will be like (sku="sku_9" and
+   * supplyChannel(id="c28c5609-9766-4176-b8d3-a72edee753e8")) or when supply channel not defined
+   * the query will be like (sku="sku_9" and supplyChannel is not defined)
+   *
+   * @param identifiers {@link Set} of unique inventory identifiers, used in search predicate
+   * @return {@link List} of inventory entry queries or empty list when there was no identifiers.
+   */
+  public static List<InventoryEntryQuery> buildQueries(
+      @Nonnull final Set<InventoryEntryIdentifier> identifiers) {
+
+    if (identifiers.isEmpty()) {
+      return emptyList();
+    }
+
+    final List<InventoryEntryQuery> inventoryEntryQueries = new ArrayList<>();
+    StringBuilder queryBuilder = new StringBuilder();
+    int limit = 0;
+    for (InventoryEntryIdentifier identifier : identifiers) {
+
+      String predicate = buildQueryPredicate(identifier).toSphereQuery();
+
+      if (queryBuilder.length() + predicate.length() > MAX_QUERY_LENGTH) {
+        inventoryEntryQueries.add(getInventoryEntryQuery(queryBuilder, limit));
```

Comment:
It seem that the limit should be counted up after `buildQueryPredicate` statement? For example if `identifiers` array has 500 SKUs and we would fit in into a single query max 250 items then it would query for 250 skus with limit of 249. Also, how do we make sure that the limit of 500 items per requests is guaranteed since otherwise we run into 409 error?

Reply:
> It seem that the limit should be counted up after buildQueryPredicate statement?

it's before because you don't want to exceed the limit when the new query is added. limit case does not seem reasonable as it's only increases when the query is added to query builder

>Also, how do we make sure that the limit of 500 items per requests is guaranteed since otherwise we run into 409 error?

size of the MAX_QUERY_LIMIT ensures a query could have max 238 sub query, the min is depend on the size of the buildQueryPredicate.
In this case the limits are based on the query and can be calculated dynamically with sizes.

https://github.com/commercetools/commercetools-sync-java/blob/fe73e1f37fc19d58f3f4bc91c2cc46d60773063c/src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryQueryBuilder.java#L17-L22

maybe the 10000/42 calculation in comment might be improved, because I figured out I'd not say in comment why it's 42, 
min size of a subquery such as: `(sku="a" and supplyChannel is not defined)` (string length = 42) but I think this could be guessed from the code comment from below as the queries will be sth like below:

https://github.com/commercetools/commercetools-sync-java/blob/fe73e1f37fc19d58f3f4bc91c2cc46d60773063c/src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryQueryBuilder.java#L28-L30